lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp

```diff
@@ -746,6 +747,110 @@ static void performNoReturnFunctionProcessing(SILModule *M,
   }
 }
 
+static bool hasRecursiveCallInPath(SILBasicBlock &Block) {
+  // Process all instructions in the block to find applies that reference
+  // the parent function.  Also looks through vtables for statically dispatched
+  // (witness) methods.
+  for (auto &I : Block) {
+    auto *AI = dyn_cast<ApplyInst>(&I);
+    if (AI && AI->getCalleeFunction()
+        && AI->getCalleeFunction() == Block.getParent())
+      return true;
+
+    if (FullApplySite FAI = FullApplySite::isa(&I)) {
+      auto &M = FAI.getModule();
```

Comment:
I would suggest to move the logic that tries to determine the target function of the `witness_method` and `class_method` into a dedicated utility function, which would return the target function if it was able to find it. This function could be useful even outside of this code snippet.

Reply:
It may also be interesting to do some checks to see if the exact dynamic type of the object, which is the operand of a `class_method`/`witness_method` instruction, is known statically. You may want to have a look at the `Devirtualize.cpp`. I think it has a lot of things you could reuse here.

Reply:
`BasicCalleeAnalysis` is definitely already doing this.  I'm going to look into refactoring this into a `BottomUpIPAnalysis` pass.